source/agora/common/crypto/Schnorr.d

```diff
@@ -0,0 +1,442 @@
+/*******************************************************************************
+
+    Low level utilities to perform Schnorr signatures on Curve25519.
+
+    Through this module, lowercase letters represent scalars and uppercase
+    letters represent points. Multiplication of a scalar by a point,
+    which is adding a point to itself multiple times, is represented with '*',
+    e.g. `a * G`. Uppercase letters are point representations of scalars,
+    that is, the scalar multipled by the generator, e.g. `r == r * G`.
+    `x` is the private key, `X` is the public key, and `H()` is the Blake2b
+    512 bits hash reduced to a scalar in the field.
+
+    Following the Schnorr BIP (see links), signatures are of the form
+    `(R,s)` and satisfy `s * G = R + H(X || R || m) * X`.
+    `r` is refered to as the nonce and is a cryptographically randomly
+    generated number that should neither be reused nor leaked.
+
+    Signature_Aggregation:
+    Since Schnorr signatures use a linear equation, they can be simply
+    combined with addition, enabling `O(1)` signature verification
+    time and `O(1)` and `O(1)` signature size.
+    Additionally, since the `c` factor does not depend on EC operation,
+    we can do batch verification, enabling us to speed up verification
+    when verifying large amount of data (e.g. a block).
+
+    See_Also:
+      - https://en.wikipedia.org/wiki/Curve25519
+      - https://en.wikipedia.org/wiki/Schnorr_signature
+      - https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki
+      - https://www.secg.org/sec1-v2.pdf
+
+    TODO:
```

Comment:
Can you turn these into issues?

Reply:
Opened #304 but the other points are not exactly "implementable" AFAICS.